### PR TITLE
Apply discounts to Outbound Shipments

### LINKED ISSUE #4703: Use different default sell prices for patient dispensing vs outbound shipments
## Is your feature request related to a problem? Please describe 👀

A requested use case is for one default sell price to be used for when shipping stock to downstream facilities and a different default sell price to be used when dispensing to patients. 

Some Stores may do both - e.g. a hospital sending stock to smaller pharmacies and also dispensing directly to patients. 

## Describe the solution you'd like 🎁

Here's the pricing workflow:

image

## Configuration
( All are items are changes to make in mSupply )
* Create a master list for all items, check the `is_default_price_list` option
* This `is_default_price_list` applies to all stores, and doesn't have to be assigned to them...
* Define prices for each item 
* Create a master list for all discounted items, set the `discount` percentage value
* This discount master list need to be assigned to any stores that should receive the discount rate.
 
## Changes required

* update master lists
  * add `is_default_price_list` to `master_list`
  * add `discount` to `master_list` 
  * and update sync and translation layers
  * add `price` to `master_list_line` and update sync & translation layers

* When stock is issued apply the `is_default_price_list` prices to the invoice line (as appropriate)
* Apply discount to outgoing invoices (apart from dispensing)

**Stocktake**
_No changes_

...OUT OF SCOPE -  POSSIBLE FUTURE REQUIREMENTS...
```
- [ ] If new stock line is introduced by stocktake, it's sell price should be set to price from `is_default_price_list`
- [ ] Stocklines are introduced from either Add Item, or Edit Item (if the stock take is created from a master list) (FRONTEND)
- [ ] Cost prices are assumed to be discounted if discount would apply to this store
- [ ] Sell Prices are not editable by the user 
```

**View Stock**
_No changes_

...OUT OF SCOPE - POSSIBLE FUTURE REQUIREMENTS...
```
- [ ] Sell Price should be prepopulated based on `is_default_price_list` once a Item is selected in the New Stock Modal
![image](https://github.com/user-attachments/assets/a5ba6642-cca7-4699-a647-5799c6168471)
- [ ] Cost prices are assumed to be discounted if discount would apply to this store
- [ ] Sell Prices are not editable by the user 
```

**Outbound Shipment**
- [x] Sell price should looked up from the `is_default_price_list`
- [x] However, a __discount__ should be applied based on the Customer's(Name) discount master list
   - [x] The global discount should apply from the masterlist to all items that are in that discount master list
   - [x] If an item is not in the discount master list no discount should be applied
- [x] Add Item (Frontend) => Modal runs graphql query to get Discount & apply to `invoice_line`)
- [x] Actions-> Allocate Lines (Backend: `allocatedOutboundShipmentUnallocatedLines`)  => Backend to look up default price list and apply price + any discount automatically
- [x] Add from Master List (Backend:`addToOutboundShipmentFromMasterList` ) => _No Changes required as pricing will be applied at allocation time_

**Inbound Shipment**
_No changes_

...OUT OF SCOPE -  POSSIBLE FUTURE REQUIREMENTS...
```
- [ ] Sell price should looked up from the `is_default_price_list`
- [ ] Discount will be applied based on the store name's discount list whether it comes via a transfer or otherwise
- [ ] Add from Master List (`addToInboundShipmentFromMasterList`) => look up `is_default_price_list` price for that item and apply it to `sell_price` 
- [ ] Cost price should come from sell price on inbound shipment
- [ ] Add Item  Frontend) => Modal runs graphql query to get`is_default_price_list` and applies it to `Sell Price`
- [ ] Cost prices are assumed to be discounted if discount would apply to this store
- [ ] Sell Prices are not editable by the user 
```

**Internal Orders**
_No changes required_

**Requsitions, R&R**
_No changes required_

**Returns**
_No changes required_

**Transfer aka in Processors code**
_No changes required_

**Dispensing**
- [x] Sell price should looked up from the `is_default_price_list` (Stockline Price is ignored)
- [x] No discounts are applied during dispensing

### Additional context 💌

Related to https://github.com/msupply-foundation/msupply/issues/15108

### Moneyworks Jobcode 🧰

`COMSCUST`

### PR DESCRIPTION
**Pull request by jmbrunskill**
Fixes #4703 (Part 3)

# 👩🏻‍💻 What does this PR do?

Changes from `discount` to `discount_percentage` to make the use of the column clearer (slightly)

Assuming a 20% discount is entered as `20` in this column.

Adds a hook to look up & apply the discount from the backend when viewing stockline to allocate

**Outbound Shipment**

- [x]  Sell price should be populated from the stockLine selected as per current process (TO BE CHANGED IN FUTURE PR)
- [x]  However, a discount should be applied based on the Customer's(Name) discount master list
- [x]  The global discount should apply from the masterlist to all items that are in that discount master list
- [x]  If an item is not in the discount master list no discount should be applied
- [x]  If the customer doesn't have a discount master list, no discount should be applied
- [x]  Add Item (Frontend) => Modal runs graphql query to get Discount & apply to invoice_line)
- [ ]  Actions-> Allocate Lines (Backend: allocatedOutboundShipmentUnallocatedLines) => Backend to look up
- [x]  Add from Master List (Backend:addToOutboundShipmentFromMasterList ) => No Changes required as pricing will be applied at allocation time

## 💌 Any notes for the reviewer?

Would be nice to show the discount in the UI somehow... Maybe a new issue?

# 🧪 Testing

- [ ] Create a master list with a `discount_percentage` (right now would need to be updated in DB until mSupply has the feature)
- [ ] Check scenarios mentioned above.

# 📃 Documentation

- [ ] **Part of an epic**: documentation will be completed for the feature as a whole
- [ ] **No documentation required**: no user facing changes or a bug fix which isn't a change in behaviour
- [X] **These areas should be updated or checked**: 
Outbound Shipment

### CODE REVIEW

File: server/service/src/pricing/mod.rs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        input: ItemDiscountLookup,
+    ) -> Result<f64, RepositoryError> {
+        get_discount_for_item_and_name_link_id(ctx, input)
+    }
+}
+
+pub struct PricingService {}
+impl PricingServiceTrait for PricingService {}
+
+fn get_default_price_list_for_store(
```
Comment (jmbrunskill) [inline server/service/src/pricing/mod.rs:23]: Will remove or move this logic before the final PR...
Just kept it here for reference for now...

File: server/graphql/general/src/queries/discount.rs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+    Ok(discount)
+}
+
+impl ItemDiscountInput {
+    pub fn to_domain(self) -> ItemDiscountLookup {
+        let ItemDiscountInput { name_id, item_id } = self;
+
+        ItemDiscountLookup {
+            name_link_id: name_id,
```
Comment (andreievg) [inline server/graphql/general/src/queries/discount.rs:47]: service layer shouldn't really be concerned with name_links, it happens in repos layer: https://github.com/msupply-foundation/open-msupply/blob/dd5ee23a3455d0ed786d68c7ae00439bd1720c5f/server/repository/src/db_diesel/master_list.rs#L93-L96

File: server/service/src/pricing/discount.rs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    let discount_master_list = match master_list {
+        Some(master_list) => master_list,
+        None => return Ok(0.0), // No discount list found, no discount
+    };
+
+    // Find if the item exists in the discount list
+    let master_list_item_line = MasterListLineRepository::new(&ctx.connection)
+        .query_by_filter(
+            MasterListLineFilter::new()
+                .master_list_id(EqualFilter::equal_to(&discount_master_list.id))
```
Comment (andreievg) [inline server/service/src/pricing/discount.rs:35]: Ideally we would add master list filter to master list line, this works though, thanks

Comment (andreievg) [reply server/service/src/pricing/discount.rs:35]: Correction haha, you need the discount from master list, nice !

**Review by andreievg — CHANGES_REQUESTED** [review]
Looks great, a small request with name_link_id please.

I think we would need to fix the dependencies to useEffect in discount, although kind of hard to judge if that useEffect will run more then once, and ideally it's all should be done with 'await' rather then other hooks and dependencies. But that's for another issue/PR

File: server/graphql/general/src/queries/discount.rs (hunk 2 above)
Comment (jmbrunskill) [reply server/graphql/general/src/queries/discount.rs]: Thanks! [d282d20](https://github.com/msupply-foundation/open-msupply/pull/4749/commits/d282d208ee44b28bd4931ad14542a78c0d7cd26c)

**Review by andreievg — APPROVED** [review]
Thanks, looks good !

**Pull request merged**